tests/suite/scale_test.go

```diff
@@ -0,0 +1,817 @@
+				}
+			}
+			Fail(fmt.Sprintf("container %s not found", name))
+			return 0
+		}
+
+		ngfRestarts := findRestarts("nginx-gateway")
+		nginxRestarts := findRestarts("nginx")
+
+		// Write results
+
+		results := scaleTestResults{
+			Name:                   testName,
+			ReloadCount:            int(reloadCount),
+			ReloadErrsCount:        int(reloadErrsCount),
+			ReloadAvgTime:          int(reloadAvgTime),
+			ReloadBuckets:          reloadBuckets,
+			EventsCount:            int(eventsCount),
+			EventsAvgTime:          int(eventsAvgTime),
+			EventsBuckets:          eventsBuckets,
+			NGFErrors:              ngfErrors,
+			NginxErrors:            nginxErrors,
+			NGFContainerRestarts:   ngfRestarts,
+			NginxContainerRestarts: nginxRestarts,
+		}
+
+		err = writeScaleResults(outFile, results)
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	runScaleResources := func(objects framework.ScaleObjects, testResultsDir string, protocol string) {
+		ttrCsvFile, writer, err := framework.NewCSVResultsWriter(testResultsDir, "ttr.csv")
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(resourceManager.Apply(objects.BaseObjects)).To(Succeed())
+
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+		defer cancel()
+		Expect(resourceManager.WaitForPodsToBeReady(ctx, ns.Name)).To(Succeed())
+
+		for i := 0; i < len(objects.ScaleIterationGroups); i++ {
+			Expect(resourceManager.Apply(objects.ScaleIterationGroups[i])).To(Succeed())
+
+			url := fmt.Sprintf("%s://%d.example.com", protocol, i)
+
+			if protocol == "http" && portFwdPort != 0 {
+				url = fmt.Sprintf("%s://%d.example.com:%d", protocol, i, portFwdPort)
+			} else if protocol == "https" && portFwdHTTPSPort != 0 {
+				url = fmt.Sprintf("%s://%d.example.com:%d", protocol, i, portFwdHTTPSPort)
+			}
+
+			startCheck := time.Now()
+
+			Eventually(
+				createResponseChecker(url, address),
+			).WithTimeout(30 * time.Second).WithPolling(100 * time.Millisecond).Should(Succeed())
+
+			ttr := time.Since(startCheck)
+
+			seconds := ttr.Seconds()
+			record := []string{strconv.Itoa(i + 1), strconv.FormatFloat(seconds, 'f', -1, 64)}
+
+			Expect(writer.Write(record)).To(Succeed())
+		}
+
+		writer.Flush()
+		Expect(ttrCsvFile.Close()).To(Succeed())
+
+		Expect(
+			framework.GenerateTTRPNG(testResultsDir, ttrCsvFile.Name(), "ttr.png"),
+		).To(Succeed())
+
+		Expect(os.Remove(ttrCsvFile.Name())).To(Succeed())
+
+		Expect(resourceManager.Delete(objects.BaseObjects)).To(Succeed())
+		for i := 0; i < len(objects.ScaleIterationGroups); i++ {
+			Expect(resourceManager.Delete(objects.ScaleIterationGroups[i])).To(Succeed())
+		}
+	}
+
+	runScaleUpstreams := func() {
+		Expect(resourceManager.ApplyFromFiles(upstreamsManifests, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+		url := "http://hello.example.com"
+		if portFwdPort != 0 {
+			url = fmt.Sprintf("http://hello.example.com:%d", portFwdPort)
+		}
```

Comment:
This structure is starting to become more common across our tests, we should probably have a function that just does it.

Reply:
agree. let's implement it separately to minimize changes in this PR

Reply:
added an issue https://github.com/nginxinc/nginx-gateway-fabric/issues/1991